ext/mbstring/libmbfl/filters/mbfilter_utf8.c

```diff
@@ -216,3 +221,106 @@ int mbfl_filt_conv_wchar_utf8(int c, mbfl_convert_filter *filter)
 
 	return 0;
 }
+
+static size_t mb_utf8_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf, size_t bufsize, unsigned int *state)
+{
+	unsigned char *p = *in, *e = p + *in_len;
+	uint32_t *out = buf, *limit = buf + bufsize;
+
+	while (p < e && out < limit) {
+		unsigned char c = *p++;
+
+		if (c < 0x80) {
+			*out++ = c;
+		} else if (c >= 0xC2 && c <= 0xDF) { /* 2 byte character */
+			if (p < e) {
+				char c2 = *p++;
+				if ((c2 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p--;
+				} else {
+					*out++ = ((c & 0x1F) << 6) | (c2 & 0x3F);
+				}
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+			}
+		} else if (c >= 0xE0 && c <= 0xEF) { /* 3 byte character */
+			if ((e - p) >= 2) {
+				char c2 = *p++;
+				char c3 = *p++;
+				if ((c2 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p -= 2;
+				} else if ((c3 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p--;
+				} else {
+					uint32_t decoded = ((c & 0xF) << 12) | ((c2 & 0x3F) << 6) | (c3 & 0x3F);
+					if (decoded < 0x800 || (decoded >= 0xD800 && decoded <= 0xDFFF)) {
+						*out++ = MBFL_BAD_INPUT;
+					} else {
+						*out++ = decoded;
+					}
+				}
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+				p = e; /* Remainder of string is too short to complete a 3-byte char; jump to end */
+			}
+		} else if (c >= 0xF0 && c <= 0xF4) { /* 4 byte character */
+			if ((e - p) >= 3) {
+				char c2 = *p++;
+				char c3 = *p++;
+				char c4 = *p++;
+				if ((c2 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p -= 3;
+				} else if ((c3 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p -= 2;
+				} else if ((c4 & 0xC0) != 0x80) {
+					*out++ = MBFL_BAD_INPUT;
+					p--;
+				} else {
+					uint32_t decoded = ((c & 0x7) << 18) | ((c2 & 0x3F) << 12) | ((c3 & 0x3F) << 6) | (c4 & 0x3F);
+					*out++ = (decoded < 0x10000) ? MBFL_BAD_INPUT : decoded;
+				}
+			} else {
+				*out++ = MBFL_BAD_INPUT;
+				p = e; /* Remainder of string is too short to complete a 4-byte char; jump to end */
+			}
+		} else {
+			*out++ = MBFL_BAD_INPUT;
+		}
+	}
+
+	*in_len = e - p;
+	*in = p;
+	return out - buf;
+}
+
+static mb_convert_buf* mb_wchar_to_utf8(uint32_t *in, size_t len, mb_convert_buf *buf, bool end)
+{
+	buf = mb_convert_buf_ensure(buf, buf->len + (len * 2));
```

Comment:
Might be a significant over-allocation for ASCII?

Reply:
Generally the allocation management in this function seems a bit fragile. Does it make much of a performance difference if the push APIs are used instead?

Reply:
Looks like the answer is... yes.
:shrug:

Reply:
Since UTF-8 is the single most common text encoding which users of this library are likely to use, if I have time, I would like to tweak this function up and see if we can make it really breathe fire. Use vector ops to peel off sequences of single-byte chars, etc.

I think allocation management will be the simplest part of the whole thing after that :laugh: